src/Cryptol/Utils/RecordMap.hs

```diff
@@ -0,0 +1,181 @@
+-- |
+-- Module      :  Cryptol.Utils.RecordMap
+-- Copyright   :  (c) 2020 Galois, Inc.
+-- License     :  BSD3
+-- Maintainer  :  cryptol@galois.com
+-- Stability   :  provisional
+-- Portability :  portable
+--
+-- This module implements an "order insensitive" datastructure for
+-- record types and values.  For most purposes, we want to deal with
+-- record fields in a canonical order; but for user interaction
+-- purposes, we generally want to display the fields in the order they
+-- were specified by the user (in source files, at the REPL, etc.).
+
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Cryptol.Utils.RecordMap
+  ( RecordMap
+  , displayOrder
```

Comment:
You should not export the field name directly, but instead define a separate top-level projection function for this. As I understand it, exporting a field name, even for an abstract datatype, allows importers of the module to use record update syntax, which can violate the datatype invariant that the module boundary is intended to preserve.

(I should probably check and see if this is still the case in recent versions of ghc.)

Reply:
I checked, and yup, ghc still lets you do this. It should be considered a bug, in my opinion.